File: javascript/tree.js
export class Node {
    constructor(key) {
        this.key = key;
        this.left = null;
        this.right = null;
    }
}

export default class BinarySearchTree{
    constructor() {
        this.root = null;
    }  

    insert(key) {
        if (this.root == null) {
            this.root = new Node(key);
        } else {
            this.insertNode(this.root, key);
        }
    }

    insertNode(node, key) {
        if (key < node.key) {
            if (node.left == null) {
                node.left = new Node(key);
            } else {
                this.insertNode(node.left, key);
            }
        } else if (key > node.key) {
            if (node.right == null) {
                node.right = new Node(key);
            } else {
                this.insertNode(node.right, key);
            }
        } else {
            return; // conflict
        }
    }

    inOrderTraverse(callback) {
        this.inOrderTraverseNode(this.root, callback);
    }

    inOrderTraverseNode(node, callback) {
        if (node != null) {
            this.inOrderTraverseNode(node.left, callback);
            callback(node.key);
            this.inOrderTraverseNode(node.right, callback);
        }       
    }

    preOrderTraverse(callback) {
        this.preOrderTraverseNode(this.root, callback);
    }

    preOrderTraverseNode(node, callback) {
        if (node != null) {
            callback(node.key);
            this.preOrderTraverseNode(node.left, callback);
            this.preOrderTraverseNode(node.right, callback);
        }       
    }

    postOrderTraverse(callback) {
        this.postOrderTraverseNode(this.root, callback);
    }

    postOrderTraverseNode(node, callback) {
        if (node != null) {
            this.postOrderTraverseNode(node.left, callback);
            this.postOrderTraverseNode(node.right, callback);
            callback(node.key);
        }       
    }

    min() {
        return this.minNode(this.root);
    }

    minNode(node) {
        let current = node;

        while (current != null && current.left != null) {
            current = current.left;
        }

        return current;
    }

    max() {
        return this.maxNode(this.root);
    }

    maxNode(node) {
        let current = node;

        while (current != null && current.right != null) {
            current = current.right;
        }

        return current;
    }

    search(key) {
        return this.searchNode(this.root, key);
    }

    searchNode(node, key) {
        if (node == null) {
            return false;
        }

        if (key < node.key) {
            return this.searchNode(node.left, key);
        } else if (key > node.key) {
            return this.searchNode(node.right, key);
        } else {
            return true;
        }
    }

    remove(key) {
        this.removeNode(this.root, key);
    }

    removeNode(node, key) {
        if (node == null) {
            return undefined;
        }
        if (key < node.key) {
            node.left = this.removeNode(node.left, key);
            return node;
        } if (key > node.key) {
            node.right = this.removeNode(node.right, key);
            return node;
        }

        // key is equal to node.item
        // handle 3 special conditions
        // 1 - a leaf node
        // 2 - a node with only 1 child
        // 3 - a node with 2 children
        // case 1
        if (node.left == null && node.right == null) {
            node = undefined;
            return node;
        }

        // case 2
        if (node.left == null) {
            node = node.right;
            return node;
        } if (node.right == null) {
            node = node.left;
            return node;
        }       

        // case 3
        const aux = this.minNode(node.right);
        node.key = aux.key;
        node.right = this.removeNode(node.right, aux.key);
        return node;
    }
}

File: python/tree.py
class Node:
    def __init__(self, key: int):
        self.key: int = key
        self.count: int = 1
        self.left: Node = None
        self.right: Node = None


class BST:
    def __init__(self):
        self.root = None

    def insert(self, key: int):
        if self.root is None:
            self.root = Node(key)
        else:
            self._insert_node(self.root, key)

    def _insert_node(self, node: Node, key: int):
        if node is None:
            node = Node(key)
            return
            
        elif key < node.key:
            if node.left is not None:
                self._insert_node(node.left, key)
            else:
                node.left = Node(key)

        elif key > node.key:
            if node.right is not None:
                self._insert_node(node.right, key)
            else:
                node.right = Node(key)

        else:
            node.count += 1  # equal/duplicated... then increment count

        

    def print_in_order(self):
        self._print_in_order_node(self.root)

    def print_pre_order(self):
        self._print_pre_order(self.root)

    def print_post_order(self):
        self._print_post_order(self.root)

    def _print_in_order_node(self, node: Node):
        if node is None:
            return  # recursion base case (stop)

        self._print_in_order_node(node.left)
        print(str(node.key) + "(" + str(node.count) + ")")
        self._print_in_order_node(node.right)

    def _print_pre_order(self, node: Node):
        if node is None:
            return  # recursion base case (stop)

        print(str(node.key) + "(" + str(node.count) + ")")
        self._print_pre_order(node.left)
        self._print_pre_order(node.right)

    def _print_post_order(self, node: Node):
        if node is None:
            return  # recursion base case (stop)

        self._print_post_order(node.left)
        self._print_post_order(node.right)
        print(str(node.key) + "(" + str(node.count) + ")")


if __name__ == "__main__":
    tree = BST()

    tree.insert(4)
    tree.insert(2)
    tree.insert(1)
    tree.insert(3)
    tree.insert(7)
    tree.insert(6)
    tree.insert(8)
    tree.insert(9)
    tree.insert(7)
    tree.insert(10)

    tree.print_in_order()
    print("------")
    tree.print_pre_order()
    print("------")
    tree.print_post_order()
